sdks/java/io/amazon-web-services/src/main/java/org/apache/beam/sdk/io/aws/sns/PublishResultCoders.java

```diff
@@ -0,0 +1,155 @@
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.apache.beam.sdk.io.aws.sns;
+
+import com.amazonaws.ResponseMetadata;
+import com.amazonaws.http.SdkHttpMetadata;
+import com.amazonaws.services.sns.model.PublishResult;
+import java.io.IOException;
+import java.io.InputStream;
+import java.io.OutputStream;
+import java.io.Serializable;
+import java.util.List;
+import org.apache.beam.sdk.coders.Coder;
+import org.apache.beam.sdk.coders.CoderException;
+import org.apache.beam.sdk.coders.StringUtf8Coder;
+import org.apache.beam.sdk.io.aws.coders.AwsCoders;
+import org.apache.beam.vendor.guava.v26_0_jre.com.google.common.collect.ImmutableList;
+
+public final class PublishResultCoders {
+
+  private PublishResultCoders() {}
+
+  public static Coder<PublishResult> defaultPublishResult() {
+    return DefaultPublishResultCoder.of();
+  }
+
+  public static Coder<PublishResult> fullPublishResult() {
+    return FullPublishResultCoder.of();
+  }
+
+  public static Coder<PublishResult> minimalPublishResult() {
+    return MinimalPublishResultCoder.of();
+  }
+
+  private static class DefaultPublishResultCoder extends Coder<PublishResult>
+      implements Serializable {
+
+    private static final DefaultPublishResultCoder INSTANCE = new DefaultPublishResultCoder();
+
+    private DefaultPublishResultCoder() {}
+
+    static DefaultPublishResultCoder of() {
+      return INSTANCE;
+    }
+
+    @Override
+    public void encode(PublishResult value, OutputStream outStream)
+        throws CoderException, IOException {
+      StringUtf8Coder.of().encode(value.getMessageId(), outStream);
+    }
+
+    @Override
+    public PublishResult decode(InputStream inStream) throws CoderException, IOException {
+      final String messageId = StringUtf8Coder.of().decode(inStream);
+      return new PublishResult().withMessageId(messageId);
+    }
+
+    @Override
+    public List<? extends Coder<?>> getCoderArguments() {
+      return ImmutableList.of();
+    }
+
+    @Override
+    public void verifyDeterministic() throws NonDeterministicException {
+      StringUtf8Coder.of().verifyDeterministic();
+    }
+  }
+
+  private abstract static class AbstractPublishResultCoder extends Coder<PublishResult>
+      implements Serializable {
+
+    private static final Coder<ResponseMetadata> RESPONSE_METADATA_CODER =
+        AwsCoders.responseMetadata();
+
+    protected abstract Coder<SdkHttpMetadata> getHttpMetadataCoder();
```

Comment:
nit: stylewise I think its simpler to understand if the abstract class has a constructor that takes the sdkhttpmetadata coder and you pass in the coder instance in the concrete subclass implementations instead of asking authors to override `getHttpMetadataCoder`